rtl/eclrtl/rtldynfield.cpp

```diff
@@ -294,6 +295,118 @@ class CRtlFieldTypeSerializer
     bool commaPending = false;
 };
 
+class CRtlFieldTypeBinSerializer
+{
+public:
+    /**
+     * Serialize a RtlTypeInfo structure to binary
+     *
+     * @param  out  Buffer for resulting serialized string
+     * @param  type RtlTypeInfo structure to be serialized
+     * @return Referenced to supplied buffer
+     */
+     static MemoryBuffer &serialize(MemoryBuffer &out, const RtlTypeInfo *type)
+     {
+         CRtlFieldTypeBinSerializer s(out);
+         s.serializeType(type);
+         return out;
+     }
+private:
+    CRtlFieldTypeBinSerializer(MemoryBuffer &_out)
+    : out(_out)
+    {
+    }
+    void serializeType(const RtlTypeInfo *type)
+    {
+        if (!serialized(type))
+        {
+            // Make sure all child types are serialized first
+            const RtlTypeInfo *child = type->queryChildType();
+            if (child)
+                serializeType(child);
+            const RtlFieldInfo * const * fields = type->queryFields();
+            if (fields)
+            {
+                for (;;)
+                {
+                    const RtlFieldInfo * child = *fields;
+                    if (!child)
+                        break;
+                    serializeType(child->type);
+                    fields++;
+                }
+            }
+            // Now serialize this one
+            types.setValue(type, nextTypeName++);
+            serializeMe(type);
+        }
+    }
+
+    void serializeMe(const RtlTypeInfo *type)
+    {
+        if (!type->canSerialize())
+            throw makeStringException(MSGAUD_user, 1000, "IFBLOCK and DICTIONARY type structures cannot be serialized");
+        unsigned fieldType = type->fieldType;
+        const char *locale = type->queryLocale();
+        if (locale && *locale)
+            fieldType |= RFTMhasLocale;
+        const RtlTypeInfo *child = type->queryChildType();
+        if (child)
+            fieldType |= RFTMhasChildType;
+        const RtlFieldInfo * const * fields = type->queryFields();
+        if (fields)
+            fieldType |= RFTMhasFields;
+        out.append(fieldType);
+        out.append(type->length);
+        if (fieldType & RFTMhasLocale)
+            out.append(locale);
+        if (child)
+            out.append(queryTypeIdx(child));
+        if (fields)
+        {
+            unsigned count = countFields(fields);
+            out.append(count);
+            for (;;)
+            {
+                const RtlFieldInfo * child = *fields;
+                if (!child)
+                    break;
+                out.append(child->name);
+                out.append(queryTypeIdx(child->type));
```

Comment:
Likely to be able to pack into 2 bytes instead of 4 - is it worth using appendPacked?  Same for count and initLength?